cluster-autoscaler/proposals/metrics.md

```diff
@@ -0,0 +1,98 @@
+## Cluster Autoscaler Monitoring
+
+## Introduction
+Currently the options to monitor Cluster Autoscaler are limited to logs, status
+configmap and events. While useful for debugging, none of this options is
+paritcularly practical for monitoring Cluster Autoscaler behaviour over time.
+This document describes a set of metrics (in Prometheus format) that will be
+added to 0.6 Cluster Autoscaler to allow better monitoring of its operations.
+
+## Current state
+Cluster Autoscaler already has a metrics endpoint providing some basic metrics.
+This includes default process metrics (number of goroutines, gc duration, cpu
+and memory details, etc) as well as some custom metrics related to time taken by
+various parts of Cluster Autoscaler main loop. Existing metrics will be renamed
+and refactored and new metrics will be added.
+
+## Metrics
+Cluster Autoscaler needs to monitor some aspects of cluster state as part of its normal operations. It can therefore provide metrics describing its own operations as well as some metrics related to general state of the cluster.
+
+All the metrics are prefixed with `cluster_autoscaler_`.
+
+### Cluster state
+
+| Metric name | Metric type | Labels | Description |
+| ----------- | ----------- | ------ | ----------- |
+| cluster_healthy | Gauge | | Whether or not cluster is healthy enough for autoscaling. 1 if it is, 0 otherwise. |
+| nodes_count | Gauge | `state`=&lt;node-state&gt; | Number of nodes in cluster. |
+| unschedulable_pods_count | Gauge | | Number of unschedulable ("Pending") pods in the cluster. |
+
+* `cluster_healthy` indicates whether cluster is healthy enough for autoscaling. CA stops all operations if significant number of nodes are unready (by default 33% as of CA 0.5.4).
+* `nodes_count` records the total number of nodes, labeled by node state. Possible
+states are `ready`, `unready`, `notStarted`.
+
+### Cluster Autoscaler execution
+This metrics are refactored from currently existing metrics and track execution
+of various parts of Cluster Autoscaler loop.
+
+| Metric name | Metric type | Labels | Description |
+| ----------- | ----------- | ------ | ----------- |
+| last_activity | Gauge | `activity`=&lt;autoscaler-activity&gt; | Last time certain part of CA logic executed |
+| duration_seconds | Summary | `function`=&lt;autoscaler-function&gt; | Time taken by various parts of CA main loop. |
```

Comment:
Just `duration_seconds` tells me little about what I would get returned by this metric. I'd suggest at least mentioning the function part so something like `function_duration_seconds`.

Also note that every label combination makes a unique time-series, so you need to make sure that values that the `function` label has is not arbitrary, but somewhat whitelisted, otherwise you may put unnecessary load on the backing time-series database.